### PR TITLE
Add secure mount function

### PR DESCRIPTION
**Pull request by leonjia0112**
1. contains unsafe libc function

### CODE REVIEW

**Comment by leonjia0112** [comment]
Added secure mount function in a separated secure mount file. Test passed, ready for a review.

File: src/crypto.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ pub fn do_hmac(
  * PBKDF2 function defaults to SHA-1 unless otherwise specified, and
  * Python-Keylime uses this default.
  */
+#[allow(dead_code)]
```
Comment (frozencemetery) [inline src/crypto.rs]: Please don't add these.

File: src/common.rs [hunk 2]
```diff
@@ -20,6 +20,15 @@ pub static IMA_ML_STUB: &'static str =
 pub static IMA_ML: &'static str =
     "/sys/kernel/security/ima/ascii_runtime_measurements";
 pub static KEY: &'static str = "secret";
+pub static MOUNT_SECURE: bool = true;
```
Comment (frozencemetery) [inline src/common.rs]: This shouldn't be a global constant.

File: src/common.rs [hunk 3]
```diff
@@ -99,6 +108,24 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
...
+ *
+ * If previledge requirement is met, change the owner of the path to root
+ * This function is unsafely using libc
+ */
+#[allow(dead_code)]
+pub fn chownroot(path: String) {
+    unsafe {
+        let euid = libc::geteuid();
+        if euid == 0 {
+            libc::chown(path.as_bytes().as_ptr() as *const i8, 0, 0);
```
Comment (frozencemetery) [inline src/common.rs]: Handle return code on this.

File: src/crypto.rs (hunk 1 above)
Comment (leonjia0112) [reply src/crypto.rs]: Without this would cause function unused warning if that is not an issue.

Comment (leonjia0112) [reply src/crypto.rs]: Got it. Thanks.

File: src/common.rs (hunk 2 above)
Comment (leonjia0112) [reply src/common.rs]: Moved to local constant.

File: src/common.rs (hunk 3 above)
Comment (leonjia0112) [reply src/common.rs]: A status code is returned to caller with execution result log.

**Comment by leonjia0112** [comment]
Test passed, ready for a review. Thanks.

File: src/common.rs [hunk 4]
```diff
@@ -99,6 +107,42 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
...
+/*
+ * Input: path directory to be changed owner to root
+ * Return: status code, 0 for successful execution, other wise fail
+ *
+ * If privilege requirement is met, change the owner of the path to root
+ * This function is unsafely using libc. Status code is return indicating
+ * execution result.
+ */
+pub fn chownroot(path: String) -> Option<i32> {
+    unsafe {
```
[the remark this answers is not in the document]
Comment (leonjia0112) [reply src/common.rs]: Thanks for your review. 

I changed the error handling in chownroot() function by return a result<> that could to be handled by the caller. The mount() function is also changed to return a result<> for error handling. Also, early return is implemented. Err is returned for error result.

The chownroot() function is actually used in multiple places in keylime_node, including functions in Common.rs and secure_moutn.rs.

Secure_mount is used for checking the mount status of the secure directory. It makes sure there is a tmpfs for keylime node to store keys, which is /var/lib/keylime/secure/ directory normally.

**Comment by leonjia0112** [comment]
Changed based on the request. Test passed, and ready for another review. Thanks.

I changed the commit message and added some inline comments for the code. Hope these could make it clearer about what the change is used for.

File: src/common.rs [hunk 5]
```diff
@@ -99,6 +108,33 @@ pub fn get_restful_parameters(urlstring: &str) -> HashMap<&str, &str> {
...
+ * If privilege requirement is met, change the owner of the path to root
+ * This function is unsafely using libc. Result is returned indicating
+ * execution result.
+ */
+pub fn chownroot(path: String) -> Result<String, i32> {
+    unsafe {
+        if libc::geteuid() != 0 {
+            error!("Privilege level unable to change ownership to root for file: {}", path);
+            return Err(-1);
+        } else {
```
Comment (frozencemetery) [inline src/common.rs]: Close.  Use more early exit  The flow should be like:

```rust
if getuid() != 0 {
    return;
}
if chown() != 0 {
    return;
}
Ok()
```

This keeps the branching manageable, which is easier to read and reason about.

**Comment by leonjia0112** [comment]
Implemented request changes. Test passed. Ready for a review. Thanks.

**Comment by frozencemetery** [comment]
Change merged.  Please take a look at the rewritten commit message.

**Pull request merged**